File: tools/figma/geometry.mjs
// 도면 좌표 ↔ 실제 렌더 좌표 대조. `npm run dev`를 띄운 상태에서:
//
//     python tools/figma/geometry.py
//     node tools/figma/geometry.mjs
//
// `audit.mjs`가 "문구가 있는가"를 세는 것과 달리 여기서는 "어디에, 얼마나 크게"를 본다.
// 화면 프레임 왼쪽 위를 원점으로 한 도면 좌표와, 문서 좌표계의 요소 위치를 맞춰본다.
//
// 판정은 **중앙값**으로 한다. 데이터가 길어지면 아래가 통째로 밀리므로 개별 어긋남보다
// 화면 전체가 어느 쪽으로 얼마나 밀렸는지가 먼저다. 세로는 데이터 양에 따라 얼마든지
// 달라지므로 가로(x)와 글자 크기를 더 무겁게 읽는다.

import fs from "fs";
import path from "path";
import { createRequire } from "module";
import { fileURLToPath } from "url";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const require = createRequire(path.join(HERE, "..", "..", "frontend", "package.json"));
const { chromium } = require("playwright");

const BASE = process.env.BASE_URL ?? "http://localhost:3000";
const geometry = JSON.parse(fs.readFileSync(path.join(HERE, "geometry.json"), "utf8"));
const ids = JSON.parse(fs.readFileSync(path.join(HERE, "audit-ids.json"), "utf8"));
const { P1, P3, M1, PICKUP, SPACE } = ids;

/** audit.mjs와 같은 경로표. 한쪽만 고치면 두 대조가 다른 화면을 본다. */
const ROUTES = {
  "C-01": ["investor", "/"],
  "C-02": ["guest", "/login"],
  "C-03": ["guest", "/signup"],
  "C-04": ["investor", "/start"],
  "C-I01": ["investor", "/verify"],
  "C-I02": ["investor", "/verify/mobile-id"],
  "C-I03": ["investor", "/verify/account"],
  "C-I05": ["investor", "/verify/done"],
  "I-01": ["investor", `/projects/${P3}`],
  "I-05": ["investor", "/investor/applications"],
  "I-06": ["investor", "/investor"],
  "I-07": ["investor", "/investor/holdings"],
  "I-09": ["investor", "/investor/notifications"],
  "I-10": ["investor", "/investor/notifications/settings"],
  "B-01": ["investor", "/subscribe"],
  "B-02": ["investor", "/subscribe/plan"],
  "B-04": ["investor", "/subscribe/order"],
  "B-06": ["investor", "/subscribe/done"],
  "B-07": ["investor", "/subscriptions"],
  "B-08": ["investor", "/subscriptions/change"],
  "B-09": ["investor", `/subscriptions/pickup/${PICKUP}`],
  "O-01": ["operator", "/operator/spaces"],
  "O-02": ["operator", `/operator/spaces/${SPACE}`],
  "O-03": ["operator", "/operator/apply"],
  "O-04": ["operator", "/operator/apply/visit"],
  "O-05": ["operator", "/operator/apply/education"],
  "O-06": ["operator", "/operator/apply/confirm"],
  "O-07": ["operator", "/operator/apply/contract"],
  "O-08": ["operator", "/operator/certificate"],
  "O-09": ["operator", "/operator"],
  "O-10": ["operator", "/operator/milestones"],
  "O-11": ["operator", `/operator/milestones/${M1}/evidence`],
  "O-11E": ["operator", `/operator/milestones/${M1}/appeal`],
  "O-13": ["operator", "/operator/settlements"],
  "A-01": ["admin", "/admin"],
  "A-02": ["admin", "/admin/operators"],
  "A-03": ["admin", "/admin/certificates"],
  "A-04": ["admin", "/admin/spaces"],
  "A-06": ["admin", "/admin/projects"],
  "A-07": ["admin", `/admin/projects/${P1}/milestones`],
  "A-08": ["admin", "/admin/evidence"],
  "A-09": ["admin", "/admin/expert-review"],
  "A-10": ["admin", "/admin/settlement-rules"],
  "A-11": ["admin", "/admin/settlements"],
  "A-12": ["admin", "/admin/audit-logs"],
  "A-13": ["admin", "/admin/roles"],
  "A-14": ["admin", "/admin/notifications"],
  "A-15": ["admin", "/admin/aml"],
  "A-16": ["admin", "/admin/ledger"],
};

const CREDS = {
  investor: "investor@farmfi.test",
  operator: "operator@farmfi.test",
  admin: "admin@farmfi.test",
};

const median = (xs) => {
  if (xs.length === 0) return null;
  const s = [...xs].sort((a, b) => a - b);
  const m = s.length >> 1;
  return s.length % 2 ? s[m] : Math.round((s[m - 1] + s[m]) / 2);
};

const browser = await chromium.launch();
const ctxs = {};
for (const [role, email] of Object.entries(CREDS)) {
  const c = await browser.newContext({ viewport: { width: 1440, height: 1000 } });
  const res = await c.request.post(`${BASE}/api/auth/login`, {
    data: { email, password: "farmfi123" },
  });
  if (!res.ok()) throw new Error(`${role} 로그인 실패 — 시드를 먼저 넣으세요`);
  ctxs[role] = c;
}
ctxs.guest = await browser.newContext({ viewport: { width: 1440, height: 1000 } });

const report = [];
for (const [sid, spec] of Object.entries(geometry)) {
  const route = ROUTES[sid];
  if (!route) continue;
  const [role, url] = route;
  // dev 서버는 첫 방문에 라우트를 컴파일한다. 한 번에 안 열리면 다시 연다 —
  // 여기서 비어 나오면 "화면이 다르다"가 아니라 "안 열렸다"인데 둘이 구분되지 않는다.
  let found = [];
  for (let attempt = 0; attempt < 2 && found.length === 0; attempt += 1) {
  const page = await ctxs[role].newPage();
  try {
    await page.goto(BASE + url, { waitUntil: "networkidle", timeout: 60000 });
    await page.waitForTimeout(attempt === 0 ? 700 : 2500);
    found = await page.evaluate((nodes) => {
      const norm = (s) => (s ?? "").replace(/[\s ]+/g, "").replace(/[’'‘]/g, "");
      // 잎 노드만 본다. 부모 컨테이너는 자식 글자를 다 품어 상자가 훨씬 크다.
      // `option`은 화면에 그려지는 상자가 없어 좌표가 0으로 나온다 — 빼지 않으면
      // 드롭다운이 있는 화면마다 수백 px씩 어긋난 것처럼 보인다.
      const SKIP = new Set(["OPTION", "SCRIPT", "STYLE", "TITLE", "NOSCRIPT"]);
      const leaves = [...document.querySelectorAll("body *")].filter((el) => {
        if (el.children.length > 0 || SKIP.has(el.tagName)) return false;
        if (norm(el.textContent).length === 0) return false;
        const r = el.getBoundingClientRect();
        // 크기가 0이면 감춰졌거나 그려지지 않은 것이다.
        return r.width > 0 && r.height > 0;
      });
      const byText = new Map();
      for (const el of leaves) {
        const k = norm(el.textContent);
        if (!byText.has(k)) byText.set(k, []);
        byText.get(k).push(el);
      }
      const out = [];
      for (const n of nodes) {
        const cands = byText.get(norm(n.text));
        if (!cands || cands.length === 0) continue;
        // 여러 개면 도면이 가리키는 자리에 가장 가까운 것을 고른다.
        let best = null;
        for (const el of cands) {
          const r = el.getBoundingClientRect();
          const x = r.left + window.scrollX;
          const y = r.top + window.scrollY;
          const d = Math.abs(x - n.x) + Math.abs(y - n.y);
          if (!best || d < best.d) {
            best = { d, x, y, size: parseFloat(getComputedStyle(el).fontSize) };
          }
        }
        out.push({
          text: n.text,
          dx: Math.round(best.x - n.x),
          dy: Math.round(best.y - n.y),
          size: Math.round(best.size),
          wantSize: n.size,
        });
      }
      return out;
    }, spec.nodes);
  } catch {
    found = [];
  }
  await page.close();
  }

  const dxs = found.map((f) => f.dx);
  const dys = found.map((f) => f.dy);
  const mx = median(dxs);
  const my = median(dys);
  // 화면 전체가 밀린 몫을 뺀 나머지가 진짜 어긋남이다.
  const offX = found.filter((f) => Math.abs(f.dx - (mx ?? 0)) > 8);
  const offY = found.filter((f) => Math.abs(f.dy - (my ?? 0)) > 24);
  const badSize = found.filter((f) => f.size !== f.wantSize);

  report.push({
    sid,
    path: url,
    matched: found.length,
    total: spec.nodes.length,
    shiftX: mx,
    shiftY: my,
    offX: offX.map((f) => `${f.text} (x${f.dx > 0 ? "+" : ""}${f.dx - (mx ?? 0)})`),
    offY: offY.map((f) => `${f.text} (y${f.dy - (my ?? 0) > 0 ? "+" : ""}${f.dy - (my ?? 0)})`),
    badSize: badSize.map((f) => `${f.text} ${f.size}px → ${f.wantSize}px`),
  });
}
await browser.close();

fs.writeFileSync(path.join(HERE, "geometry-report.json"), JSON.stringify(report, null, 2));

console.log("화면    맞춘문구  가로밀림  세로밀림  가로어긋남  세로어긋남  글자크기");
const sorted = [...report].sort(
  (a, b) => b.offX.length + b.badSize.length - (a.offX.length + a.badSize.length),
);
for (const r of sorted) {
  console.log(
    [
      r.sid.padEnd(6),
      `${r.matched}/${r.total}`.padStart(8),
      String(r.shiftX ?? "-").padStart(8),
      String(r.shiftY ?? "-").padStart(8),
      String(r.offX.length).padStart(10),
      String(r.offY.length).padStart(10),
      String(r.badSize.length).padStart(8),
    ].join(" "),
  );
}
const sum = (k) => report.reduce((n, r) => n + r[k].length, 0);
console.log(
  `\n합계 — 가로 어긋남 ${sum("offX")} · 세로 어긋남 ${sum("offY")} · 글자 크기 ${sum("badSize")}` +
    ` (맞춘 문구 ${report.reduce((n, r) => n + r.matched, 0)})`,
);

File: tools/figma/geometry.py
"""덤프에서 화면별 '문구 + 좌표 + 글자 크기'를 뽑아 `tools/figma/geometry.json`에 쓴다.

    python tools/figma/geometry.py

`labels.py`가 "문구가 있는가"를 보는 것과 달리 여기서는 "어디에, 얼마나 크게"를 담는다.
좌표는 화면 프레임 왼쪽 위를 원점으로 한 상대값이다 — 캔버스 절대좌표는 화면마다
달라 브라우저와 맞출 수 없다.
"""
import json
import os
import re
import sys

ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
HERE = os.path.dirname(os.path.abspath(__file__))
BASE = os.path.join(ROOT, "design", "screens", "farmfi-web")
OUT = os.path.join(HERE, "geometry.json")

sys.path.insert(0, HERE)
from labels import SCREENS  # noqa: E402

BOX = re.compile(r"@(-?\d+),(-?\d+)\s+(\d+)x(\d+)")
TEXT = re.compile(r"text='(.*?)'\s+(\S+)\s+([A-Za-z ]+?)\s+(\d+)px")


def nodes(path):
    """(원점, 텍스트 노드 목록). 첫 줄이 화면 프레임이고 그게 원점이다."""
    lines = open(path, encoding="utf-8").read().splitlines()
    if not lines:
        return None, []
    head = BOX.search(lines[0])
    if not head:
        return None, []
    ox, oy = int(head.group(1)), int(head.group(2))
    frame_w, frame_h = int(head.group(3)), int(head.group(4))

    out = []
    for line in lines:
        m = TEXT.search(line)
        b = BOX.search(line)
        if not m or not b:
            continue
        s = m.group(1).replace("\\n", " ").replace(" ", " ").strip()
        # 숫자가 든 문구는 데이터라 화면마다 값이 달라 위치 대조에 못 쓴다.
        if not s or len(s) < 2 or re.search(r"\d", s):
            continue
        out.append(
            {
                "text": s,
                "x": int(b.group(1)) - ox,
                "y": int(b.group(2)) - oy,
                "w": int(b.group(3)),
                "h": int(b.group(4)),
                "size": int(m.group(4)),
                "weight": m.group(3).strip(),
            }
        )

    # 같은 문구가 여러 번 나오면 좌표를 하나로 정할 수 없다. 빼고 센다.
    seen = {}
    for n in out:
        seen[n["text"]] = seen.get(n["text"], 0) + 1
    return {"w": frame_w, "h": frame_h}, [n for n in out if seen[n["text"]] == 1]


def pick(rel):
    """같은 화면이 두 프레임으로 갈린 경우(모달·배경) 1440 폭 쪽이 화면 전체다.

    카드만 담긴 프레임을 쓰면 원점이 카드 왼쪽 위가 되어, 화면 안에서 카드가
    어디 놓였는지가 통째로 빠진다 — 좌표 대조가 그만큼 어긋난다.
    """
    base = os.path.join(BASE, rel)
    alt = base[:-4] + "-2.txt"
    for path in (alt, base):
        if not os.path.exists(path):
            continue
        frame, ns = nodes(path)
        if frame and frame["w"] == 1440 and ns:
            return path
    return base


def main():
    data = {}
    for sid, rel in SCREENS.items():
        path = pick(rel)
        if not os.path.exists(path):
            continue
        frame, ns = nodes(path)
        if frame is None or not ns:
            continue
        data[sid] = {"frame": frame, "nodes": ns, "dump": os.path.relpath(path, BASE).replace("\\", "/")}
    with open(OUT, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=0)
    print(f"화면 {len(data)} · 좌표 있는 문구 {sum(len(v['nodes']) for v in data.values())} → {OUT}")


if __name__ == "__main__":
    main()
